src/common/src/array/mod.rs

```diff
@@ -483,14 +483,9 @@ macro_rules! impl_array_builder {
                     None => match self {
                         $( Self::$variant_name(inner) => inner.append_n(n, None), )*
                     }
-                    Some(scalar_ref) => match (self, scalar_ref) {
-                        $( (Self::$variant_name(inner), ScalarRefImpl::$variant_name(v)) => inner.append_n(n, Some(v)), )*
-                        (this_builder, this_scalar_ref) => panic!(
-                            "Failed to append datum, array builder type: {}, scalar type: {}",
-                            this_builder.get_ident(),
-                            this_scalar_ref.get_ident()
-                        ),
-                    },
+                    Some(scalar_ref) => paste::paste! { match self {
+                        $( Self::$variant_name(inner) => inner.append_n(n, Some(scalar_ref.[<into_ $suffix_name>]())), )*
```

Comment:
Does `into` also result in `n` lookup?

Reply:
I'm not sure about the compiler's optimization. 🥺 Let's wait for the flamegraph.